base_onchange_rule/models/base_onchange_rule.py

```diff
@@ -0,0 +1,95 @@
+    _description = "Base Onchange Rule"
+    _order = "sequence"
+
+    active = fields.Boolean(default=False, tracking=True)
+    name = fields.Char(string="Rule Name", required=True)
+    description = fields.Text()
+    model = fields.Many2one(
+        "ir.model", domain=lambda r: [("model", "!=", r._name)], required=True
+    )
+    model_name = fields.Char(related="model.model")
+    code = fields.Text(
+        "Python Code",
+        default=DEFAULT_PYTHON_CODE,
+    )
+    sequence = fields.Integer(string="Sequence", required=True, default=10)
+    company_id = fields.Many2one(
+        "res.company",
+        string="Company",
+        default=lambda self: self.env.company,
+    )
+    fields_lines = fields.One2many(
+        "ir.server.object.lines",
+        "base_onchange_rule_id",
+        string="Value Mapping",
+        copy=True,
+    )
+    onchange_fields = fields.Many2many("ir.model.fields")
+
+    def evaluate_rules(self, records):
+        for rule in self:
+            company = rule.company_id or self.env.company
+            records = records.with_context(company_id=company.id)
+            # 1. Filter records matching the same company
+            if rule.company_id:
+                company_records = records.filtered(lambda r: company == rule.company_id)
+            else:
+                company_records = records
+            # 2. Filter records matching the onchange conditions
+            records_matching = rule._evaluate_py_code(company_records)
+            if not records_matching:
+                continue
+            # 3. Get values to be written
+            res = {}
+            for exp in rule.fields_lines:
+                res[exp.col1.name] = exp.eval_value()[exp.id]
+            # 4. Update values
+            records_matching.update(res)
+
+    def _evaluate_py_code(self, records):
+        self.ensure_one()
+        expr = self.code
+        space = {"records": records, "model": self.env[records._name]}
+        try:
+            safe_eval(expr, space, mode="exec", nocopy=True)
+        except Exception as e:
+            raise UserError(_("Error when the code in:\n %s \n\n(%s)") % (self.name, e))
+        return space.get("res", False)
+
+    @api.onchange("model")
+    def _onchange_model(self):
+        for rec in self:
+            rec.fields_lines.write({"onchange_rule_model_name": rec.model.model})
+
+
+class IrServerObjectLines(models.Model):
+    _inherit = "ir.server.object.lines"
+
+    base_onchange_rule_id = fields.Many2one(
+        "base.onchange.rule",
+        string="Related Base Onchange Rule",
+        ondelete="cascade",
+    )
+    onchange_rule_model_name = fields.Char(
+        compute="_compute_onchange_rule_model_name", store=True
+    )
+
+    @api.depends("base_onchange_rule_id", "base_onchange_rule_id.model")
+    def _compute_onchange_rule_model_name(self):
+        for rec in self:
+            rec.onchange_rule_model_name = rec.base_onchange_rule_id.model_name
```

Comment:
```suggestion
        related="base_onchange_rule_id.model_name", store=True
    )
```

Reply:
When the record has still not been created the related values can't be assigned

Reply:
makes sense 👍🏽